src/linux_ip_socket/linux_ip_socket.cpp

```diff
@@ -0,0 +1,336 @@
+
+    const std::string remote_port = std::to_string(port);
+
+    const int getaddrinfo_result = getaddrinfo(address, remote_port.c_str(), &hints, target);
+
+    if(getaddrinfo_result != 0) {
+        std::cerr << "getaddrinfo returned an error: " << gai_strerror(getaddrinfo_result) << std::endl;
+        std::cerr << "Aborting." << std::endl;
+        abort();
+    }
+}
+
+void
+linux_ip_socket_private_data::parse_recv_buffer(const size_t length)
+{
+    for(size_t i = 0; i < length; ++i) {
+        switch(m_parse_state) {
+            case STATE_WAIT:
+                if(m_recv_buffer[i] == START_BYTE) {
+                    m_parse_state = STATE_DATA_BYTE;
+                }
+                break;
+            case STATE_DATA_BYTE:
+                if(m_recv_buffer[i] == STOP_BYTE) {
+                    Broker_receive_packet(m_message_buffer, m_message_buffer_index);
+                    m_message_buffer_index = 0;
+                    m_parse_state = STATE_WAIT;
+                } else if(m_recv_buffer[i] == ESCAPE_BYTE) {
+                    m_parse_state = STATE_ESCAPE_BYTE;
+                } else if(m_recv_buffer[i] == START_BYTE) {
+                    m_message_buffer_index = 0;
+                    m_parse_state = STATE_DATA_BYTE;
+                } else {
+                    m_message_buffer[m_message_buffer_index] = m_recv_buffer[i];
+                    ++m_message_buffer_index;
+                }
+                break;
+            case STATE_ESCAPE_BYTE:
+                m_message_buffer[m_message_buffer_index] = m_recv_buffer[i];
+                ++m_message_buffer_index;
+                m_parse_state = STATE_DATA_BYTE;
+                break;
+        }
+    }
+}
+
+void
+linux_ip_socket_private_data::send_packet(const int sockfd, const size_t buffer_length)
+{
+    const int send_result = send(sockfd, m_send_buffer, buffer_length, 0);
+    if(send_result == -1) {
+        std::cerr << "sendto error\n";
+        return;
+    }
+}
+
+int
+linux_ip_socket_private_data::connect_to_remote_driver()
+{
+    addrinfo* address_array = nullptr;
+
+    find_addresses(&address_array, m_ip_remote_device_configuration->address, m_ip_remote_device_configuration->port);
+
+    addrinfo* connect_address = address_array;
+
+    if(connect_address == nullptr) {
+        std::cerr << "Cannot find remote address." << std::endl;
+        return INVALID_SOCKET_ID;
+    }
+
+    const int sockfd = socket(connect_address->ai_family, connect_address->ai_socktype, connect_address->ai_protocol);
+    if(sockfd == -1) {
+        std::cerr << "socket() returned an error: " << strerror(errno) << std::endl;
+        freeaddrinfo(address_array);
+        return INVALID_SOCKET_ID;
+    }
+    const int connect_result = connect(sockfd, connect_address->ai_addr, connect_address->ai_addrlen);
+    if(connect_result == -1) {
+        std::cerr << "connect() returned an error: " << strerror(errno) << std::endl;
+        freeaddrinfo(address_array);
```

Comment:
What do you think about doing try {} finally { freeaddrinfo(address_array)}? Just a suggestion.

Reply:
In C++ there's no `finally`. But I can refactor it to call `freeaddrinfo` before checking `connect_result`